contracts/src/CrossChainApplications/NativeTokenBridge/NativeTokenDestination.sol

```diff
@@ -0,0 +1,207 @@
+            } else {
+                currentReserveImbalance -= amount;
+                emit CollateralAdded({amount: amount, remaining: currentReserveImbalance});
+                return;
+            }
+        }
+
+        // Calls NativeMinter precompile through INativeMinter interface.
+        _nativeMinter.mintNativeCoin(recipient, adjustedAmount);
+        totalMinted += adjustedAmount;
+        emit NativeTokensMinted(recipient, adjustedAmount);
+    }
+
+    /**
+     * @dev See {INativeTokenDestination-transferToSource}.
+     */
+    function transferToSource(
+        address recipient,
+        TeleporterFeeInfo calldata feeInfo,
+        address[] calldata allowedRelayerAddresses
+    ) external payable nonReentrant {
+        // The recipient cannot be the zero address.
+        require(
+            recipient != address(0),
+            "NativeTokenDestination: zero Recipient Address"
+        );
+
+        require(
+            currentReserveImbalance == 0,
+            "NativeTokenDestination: Cannot release tokens until source contract is collateralized"
+        );
+
+        // Lock tokens in this bridge instance. Supports "fee/burn on transfer" ERC20 token
+        // implementations by only bridging the actual balance increase reflected by the call
+        // to transferFrom.
+        uint256 adjustedFeeAmount = 0;
+        if (feeInfo.amount > 0) {
+            adjustedFeeAmount = SafeERC20TransferFrom.safeTransferFrom(
+                IERC20(feeInfo.contractAddress),
+                feeInfo.amount
+            );
+            IERC20(feeInfo.contractAddress).safeIncreaseAllowance(
+                address(teleporterMessenger),
+                adjustedFeeAmount
+            );
+        }
+
+        // Burn native token by sending to BLACKHOLE_ADDRESS
+        payable(BLACKHOLE_ADDRESS).transfer(msg.value);
+
+        uint256 messageID = teleporterMessenger.sendCrossChainMessage(
+            TeleporterMessageInput({
+                destinationChainID: sourceBlockchainID,
+                destinationAddress: nativeTokenSourceAddress,
+                feeInfo: feeInfo,
+                requiredGasLimit: TRANSFER_NATIVE_TOKENS_REQUIRED_GAS,
+                allowedRelayerAddresses: allowedRelayerAddresses,
+                message: abi.encode(recipient, msg.value)
+            })
+        );
+
+        emit TransferToSource({
+            sender: msg.sender,
+            recipient: recipient,
+            amount: msg.value,
+            teleporterMessageID: messageID
+        });
+    }
+
+    function isCollateralized() external view returns (bool) {
+        return currentReserveImbalance == 0;
+    }
+
+    function totalSupply() external view returns (uint256) {
+        uint256 burned = address(BURNED_TX_FEES_ADDRESS).balance - address(BLACKHOLE_ADDRESS).balance;
+        require(burned > totalMinted + initialReserveImbalance, "NativeTokenDestination: FATAL - Contract has tokens unaccounted for");
```

Comment:
This is an interesting way to broadcast this fatal condition... We should check this before adjusting the supply elsewhere of course. I'm not sure that reverting this function call is the right way to do this. It might be better to defer to the contract caller to validate the total supply. A separate monitoring application, for instance.

Reply:
This `require` is really just solving the problem of underflow here. if `burned > totalMinted + initialReserveImbalance`, then the value on return would underflow. I don't really expect this condition to ever happen, but we should just check it to make sure that we're never returning a junk number. I don't think it's worth any extra complexity to check this elsewhere.

Reply:
I see, thanks for clarifying